### PR TITLE
fs: cluster_allocator: added method returning future waiting for clusters

### PR DESCRIPTION
**Pull request by tropuq**

### CODE REVIEW

File: src/fs/cluster_allocator.hh [hunk 1]
```diff
@@ -22,29 +22,45 @@
...
 public:
-    explicit cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters);
+    cluster_allocator();
+
+    cluster_allocator(const cluster_allocator&) = delete;
+    cluster_allocator& operator=(const cluster_allocator&) = delete;
+    cluster_allocator(cluster_allocator&&) = default;
+    cluster_allocator& operator=(cluster_allocator&&) = delete;
+
+    void init(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters);
```
Comment (varqox) [inline src/fs/cluster_allocator.hh]: why init instead of constructor?

File: src/fs/cluster_allocator.hh [hunk 2]
```diff
@@ -22,29 +22,45 @@
...
 
     // Tries to allocate a cluster
     std::optional<cluster_id_t> alloc();
 
+    // Waits for free @p count clusters and allocates
+    future<std::vector<cluster_id_t>> alloc_wait(size_t count = 1);
+
     // @p cluster_id has to be allocated using alloc()
```
Comment (varqox) [inline src/fs/cluster_allocator.hh]:
```suggestion
    // @p cluster_id has to be allocated using alloc() or alloc_wait()
```

File: src/fs/cluster_allocator.hh [hunk 3]
```diff
@@ -22,29 +22,45 @@
...
     void free(cluster_id_t cluster_id);
-};
 
-}
+    // @p cluster_id has to be allocated using alloc()
```
Comment (varqox) [inline src/fs/cluster_allocator.hh]:
```suggestion
    // @p cluster_id has to be allocated using alloc() or alloc_wait()
```

File: src/fs/cluster_allocator.cc [hunk 4]
```diff
@@ -21,34 +21,71 @@
...
+    _cluster_sem.consume(1);
+
+    return do_alloc();
+}
+
+future<std::vector<cluster_id_t>> cluster_allocator::alloc_wait(size_t count) {
+    return _cluster_sem.wait(count).then([this, count] {
+        std::vector<cluster_id_t> cluster_ids;
+        for (size_t i = 0; i < count; ++i) {
+            cluster_ids.emplace_back(do_alloc());
```
Comment (varqox) [inline src/fs/cluster_allocator.cc:90]: If this fails we need to rollback previous allocations to provide strong exception guarantee, as it is cheap here

File: src/fs/cluster_allocator.cc [hunk 5]
```diff
@@ -21,34 +21,71 @@
...
 
     cluster_id_t cluster_id = _free_clusters.front();
     _free_clusters.pop_front();
     _allocated_clusters.insert(cluster_id);
+
     return cluster_id;
 }
 
-void cluster_allocator::free(cluster_id_t cluster_id) {
+cluster_id_t cluster_allocator::do_free(cluster_id_t cluster_id) {
```
Comment (varqox) [inline src/fs/cluster_allocator.cc]: 1. returning `cluster_id_t` looks pointless.
2. A comment could indicate that the strong exception guarantee is provided

File: src/fs/cluster_allocator.cc [hunk 6]
```diff
@@ -21,34 +21,71 @@
...
-std::optional<cluster_id_t> cluster_allocator::alloc() {
-    if (_free_clusters.empty()) {
-        return std::nullopt;
-    }
+cluster_id_t cluster_allocator::do_alloc() {
+    assert(not _free_clusters.empty());
 
     cluster_id_t cluster_id = _free_clusters.front();
     _free_clusters.pop_front();
     _allocated_clusters.insert(cluster_id);
```
Comment (varqox) [inline src/fs/cluster_allocator.cc]: If it throws, there is no strong exception guarantee -- fix that

File: src/fs/cluster_allocator.cc [hunk 7]
```diff
@@ -21,34 +21,71 @@
...
+cluster_allocator::cluster_allocator() : _cluster_sem(0) {}
 
-cluster_allocator::cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters)
-        : _allocated_clusters(std::move(allocated_clusters)), _free_clusters(std::move(free_clusters)) {}
+void cluster_allocator::init(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters) {
+    _allocated_clusters = std::move(allocated_clusters);
+    _free_clusters = std::move(free_clusters);
+    assert(_cluster_sem.waiters() == 0);
+    _cluster_sem.consume(_cluster_sem.available_units());
+    _cluster_sem.signal(_free_clusters.size());
```
Comment (varqox) [inline src/fs/cluster_allocator.cc:55]: something like `_cluster_sem = semaphore(_free_clusters.size())` doesn't do the job?

File: src/fs/cluster_allocator.cc [hunk 8]
```diff
@@ -21,34 +21,71 @@
...
 }
 
+void cluster_allocator::free(cluster_id_t cluster_id) {
+    do_free(cluster_id);
+    _cluster_sem.signal();
 }
+
+void cluster_allocator::free(const std::vector<cluster_id_t>& cluster_ids) {
+    for (auto& cluster_id : cluster_ids) {
+        do_free(cluster_id);
```
Comment (varqox) [inline src/fs/cluster_allocator.cc:103]: If this throws, strong exception guarantee is broken -- fix that.

File: src/fs/cluster_allocator.hh [hunk 9]
```diff
@@ -22,29 +22,45 @@
...
+    cluster_allocator& operator=(const cluster_allocator&) = delete;
+    cluster_allocator(cluster_allocator&&) = default;
+    cluster_allocator& operator=(cluster_allocator&&) = delete;
+
+    void init(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters);
 
     // Tries to allocate a cluster
     std::optional<cluster_id_t> alloc();
 
+    // Waits for free @p count clusters and allocates
```
Comment (varqox) [inline src/fs/cluster_allocator.hh]:
```suggestion
    // Waits until @p count free clusters are available and allocates them
```

**Review by varqox — CHANGES_REQUESTED** [review]

File: src/fs/cluster_allocator.cc (hunk 7 above)
Comment (tropuq) [reply src/fs/cluster_allocator.cc:55]: Assignment operators are deleted for semaphore.

Comment (varqox) [reply src/fs/cluster_allocator.cc:55]: Oh, so be it then.
Btw. I've an idea that we could assert that available_units() == _free_clusters.size() at the beginning of the function to make sure that nobody is using the allocator right now.

File: src/fs/cluster_allocator.cc [hunk 10]
```diff
@@ -21,34 +21,86 @@
...
+    free_clusters.reserve(cluster_nb);
+
+    _allocated_clusters = std::move(allocated_clusters);
+    _free_clusters = std::move(free_clusters);
+    _cluster_sem.consume(_cluster_sem.available_units());
+    _cluster_sem.signal(_free_clusters.size());
+}
+
+cluster_id_t cluster_allocator::do_alloc() noexcept {
+    assert(not _free_clusters.empty());
```
Comment (varqox) [inline src/fs/cluster_allocator.cc]:
```suggestion
    assert(!_free_clusters.empty());
```

Comment (varqox) [reply src/fs/cluster_allocator.cc]: People do not like `not`, `and` and `or` unfortunately :(

File: src/fs/cluster_allocator.cc [hunk 11]
```diff
@@ -21,34 +21,86 @@
...
 
     cluster_id_t cluster_id = _free_clusters.front();
     _free_clusters.pop_front();
-    _allocated_clusters.insert(cluster_id);
+    _allocated_clusters.emplace(cluster_id);
```
Comment (varqox) [inline src/fs/cluster_allocator.cc]: I am not sure that reserve on unordered_set makes emplace noexcept. Documentations says that it only reserves buckets, but I don't know what about allocating the list element on a hash collision (or even when there is no collision, but this is the first element in that bucket).

File: src/fs/cluster_allocator.hh [hunk 12]
```diff
@@ -22,29 +22,47 @@
...
+    cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters = {},
+            circular_buffer<cluster_id_t> free_clusters = {});
+
+    cluster_allocator(const cluster_allocator&) = delete;
+    cluster_allocator& operator=(const cluster_allocator&) = delete;
+    cluster_allocator(cluster_allocator&&) = default;
+    cluster_allocator& operator=(cluster_allocator&&) = delete;
+
+    // Changes cluster_allocator's set of free and allocated clusters. Assumes that there are no allocated clusters
+    // before call.
```
Comment (varqox) [inline src/fs/cluster_allocator.hh]:
```suggestion
    // Changes cluster_allocator's set of free and allocated clusters. Assumes that there are no allocated clusters
    // and nobody uses the cluster_allocator (e.g. waiting to alloc() a cluster).
```

File: tests/unit/fs_cluster_allocator_test.cc [hunk 13]
```diff
@@ -19,19 +19,46 @@
  * Copyright (C) 2019 ScyllaDB
  */
 
+#include "fs/cluster.hh"
+#include "seastar/core/circular_buffer.hh"
 #define BOOST_TEST_MODULE fs
 
 #include "fs/cluster_allocator.hh"
```
Comment (varqox) [inline tests/unit/fs_cluster_allocator_test.cc]:
```suggestion
#define BOOST_TEST_MODULE fs

#include "fs/cluster.hh"
#include "fs/cluster_allocator.hh"
#include "seastar/core/circular_buffer.hh"
```

**Review by varqox — CHANGES_REQUESTED** [review]

File: src/fs/cluster_allocator.cc (hunk 11 above)
Comment (tropuq) [reply src/fs/cluster_allocator.cc]: https://en.cppreference.com/w/cpp/container/unordered_set
Here it says that reserve "reserves space for at least the specified number of elements. This regenerates the hash table." And for rehash "reserves at least the specified number of buckets. This regenerates the hash table." But that is the only place where I found it.

Comment (tropuq) [reply src/fs/cluster_allocator.cc]: You are right. I found that question.
https://stackoverflow.com/questions/42373232/why-does-stdunordered-map-have-a-reserve-method

File: src/fs/cluster_allocator.hh [hunk 14]
```diff
@@ -22,29 +22,49 @@
 #pragma once
 
 #include "fs/cluster.hh"
+#include "seastar/core/circular_buffer.hh"
+#include "seastar/core/future.hh"
+#include "seastar/core/semaphore.hh"
 
-#include <deque>
 #include <optional>
 #include <unordered_set>
```
Comment (varqox) [inline src/fs/cluster_allocator.hh:32]:
```suggestion
#include <unordered_map>
```

File: src/fs/cluster_allocator.cc [hunk 15]
```diff
@@ -21,34 +21,98 @@
...
+        circular_buffer<cluster_id_t> free_clusters) {
+    assert((size_t)_cluster_sem.available_units() == _free_clusters.size());
+    assert(_cluster_sem.waiters() == 0);
+
+    free_clusters.reserve(free_clusters.size() + allocated_clusters.size());
+    _allocated_clusters = create_allocated_clusters(allocated_clusters, free_clusters);
+    _free_clusters = std::move(free_clusters);
+
+    _cluster_sem.consume(_cluster_sem.available_units());
+    _cluster_sem.signal(_free_clusters.size());
```
Comment (varqox) [inline src/fs/cluster_allocator.cc]: maybe calling reset() from a constructor will be sufficient? I.e. constructor calls default constructor and then calls reset(), then you could inline create_allocated_clusters in the reset().

File: src/fs/cluster_allocator.hh [hunk 16]
```diff
@@ -22,29 +22,49 @@
...
+    circular_buffer<cluster_id_t> _free_clusters;
+    semaphore _cluster_sem;
+
+    cluster_id_t do_alloc() noexcept;
+    void do_free(cluster_id_t cluster_id) noexcept;
 
 public:
-    explicit cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters);
+    cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters = {},
+            circular_buffer<cluster_id_t> free_clusters = {});
```
Comment (varqox) [inline src/fs/cluster_allocator.hh]: it is not OK, that we have here one argument constructor that is also viable for implicit conversions... Please add default constructor and remove default arguments here.

File: src/fs/cluster_allocator.hh [hunk 17]
```diff
@@ -22,29 +22,49 @@
...
+    cluster_allocator& operator=(const cluster_allocator&) = delete;
+    cluster_allocator(cluster_allocator&&) = default;
+    cluster_allocator& operator=(cluster_allocator&&) = delete;
+
+    // Changes cluster_allocator's set of free and allocated clusters. Assumes that there are no allocated clusters
+    // and nobody uses the cluster_allocator (e.g. waiting to alloc() a cluster).
+    // Strong exception guarantee is provided.
+    void reset(std::unordered_set<cluster_id_t> allocated_clusters, circular_buffer<cluster_id_t> free_clusters);
 
     // Tries to allocate a cluster
```
Comment (varqox) [inline src/fs/cluster_allocator.hh]:
```suggestion
    // Tries to allocate a cluster.
    // Strong exception guarantee is provided.
```

File: tests/unit/fs_cluster_allocator_test.cc [hunk 18]
```diff
@@ -20,18 +20,44 @@
  */
 
 #define BOOST_TEST_MODULE fs
```
Comment (varqox) [inline tests/unit/fs_cluster_allocator_test.cc:22]:
```suggestion
#define BOOST_TEST_MODULE fs

```

**Review by varqox — APPROVED** [review]

**Review by rokinsky — APPROVED** [review]

**Pull request merged**